socorro/external/fs/crashstorage.py

```diff
@@ -663,68 +659,115 @@ class TarFileCrashStore(CrashStorageBase):
     @staticmethod
     def stringify_datetimes(obj):
         if isinstance(obj, datetime.datetime):
-            return obj.strftime("%Y-%m-%d %H:%M:%S.%f")
-        raise TypeError
+            return obj.strftime("%Y-%m-%dT%H:%M:%S.%f")
+        return json.JSONEncoder.default(self, obj)
+
+    #------------------------------------------------------------------------------
+    def _create_tarfile(self):
+        """subclasses that have a different way of openning or creating
+        the tar file pointer can override this method.  Useful for creating
+        text buffer tarfiles or using temporary files"""
+        return self.tarfile_module.open(self.config.tarball_name, 'w')
 
+    #------------------------------------------------------------------------------
     def __init__(self, config, quit_check_callback=None):
-        super(TarFileCrashStore, self).__init__(config, quit_check_callback)
+        super(TarFileWritingCrashStore, self).__init__(config, quit_check_callback)
         self.tarfile_module = config.tarfile_module
         self.gzip_module = config.gzip_module
-        self.os_module = config.os_module
-
-    def _save_to_tarfile(self, actual_pathname, target_pathname):
-        try:
-            self.tar_file.add(actual_pathname, target_pathname)
-        except AttributeError:
-            # the tar_file is lazy instantiated.  It isn't created until
-            # a process tries to save something to it
-            self.tar_file = self.tarfile_module.open(
-                self.config.tarball_name,
-                'w'
-            )
-            self.tar_file.add(actual_pathname, target_pathname)
+        self.tar_fp = self._create_tarfile()
 
+    #------------------------------------------------------------------------------
     def close(self):
-        try:
-            self.tar_file.close()
-        except AttributeError:
-            # the tar_file was never actually created because the save_*
-            # were never called.  we can silently ignore this
-            pass
+        self.tar_fp.close()
 
+    #------------------------------------------------------------------------------
     def save_processed(self, processed_crash):
         processed_crash_as_string = json.dumps(
             processed_crash,
             default=self.stringify_datetimes
         )
-        crash_id = processed_crash['crash_id']
-        file_name = os.path.join(
-            self.config.temp_directory,
-            crash_id + '.jsonz'
+        crash_id = processed_crash["crash_id"]
+
+        compressed_crash = StringIO()
+        gzip_file = self.gzip_module.GzipFile(fileobj=compressed_crash, mode='w')
+        gzip_file.write(processed_crash_as_string)
+        gzip_file.close()
+        compressed_crash.seek(0)
+        tarinfo = self.tarfile_module.TarInfo('%s.jsonz' % crash_id)
+        tarinfo.size = len(compressed_crash.getvalue())
+        self.tar_fp.addfile(tarinfo, compressed_crash)
```

Comment:
Uh?! If you've read the StringIO thing to get its length (line above), how come you don't have to `seek(0)` on it again?